### PR TITLE
Sync breadcrumbs and documentation panel.

### PR DESCRIPTION
**Pull request by MichaelMauderer**
### Pull Request Description

Implements #7310.

Peek 2023-08-09 16-07

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: lib/rust/ensogl/component/breadcrumbs/src/entry.rs [hunk 1]
```diff
@@ -186,8 +186,9 @@ impl EntryData {
         let separator = separator::View::new();
         let state = default();
         let icon: any_icon::View = default();
-        icon.set_size((ICON_WIDTH, ICON_WIDTH));
+        ellipsis.set_size((ICON_WIDTH, ICON_WIDTH));
```
Comment (vitvakatu) [inline lib/rust/ensogl/component/breadcrumbs/src/entry.rs]: The next line sets the size of the ellipsis as well.

File: lib/rust/ensogl/component/breadcrumbs/src/entry.rs [hunk 2]
```diff
@@ -263,8 +263,8 @@ impl EntryData {
         self.separator.set_size(Vector2(separator::ICON_WIDTH, size.y));
         self.ellipsis.set_size(Vector2(ellipsis::ICON_WIDTH, size.y));
         self.icon.set_size(Vector2(ICON_WIDTH, size.y));
-        self.icon.set_x(-ICON_WIDTH);
-        self.icon.set_y(-2.0);
+        self.icon.set_x(-size.x / 2.0 - 2.0);
+        self.icon.set_y(-ICON_WIDTH * 1.25 - 2.0);
```
Comment (vitvakatu) [inline lib/rust/ensogl/component/breadcrumbs/src/entry.rs]: Offsets like `2.0` and `1.25` should probably be moved to constants.

File: app/gui/view/component-browser/component-list-panel/Cargo.toml [hunk 3]
```diff
@@ -18,6 +18,7 @@ ensogl-selector = { path = "../../../../../lib/rust/ensogl/component/selector" }
 ensogl-toggle-button = { path = "../../../../../lib/rust/ensogl/component/toggle-button" }
 ensogl-text = { path = "../../../../../lib/rust/ensogl/component/text" }
 ensogl-tooltip = { path = "../../../../../lib/rust/ensogl/component/tooltip/" }
+ensogl-breadcrumbs = { path = "../../../../../lib/rust/ensogl/component/breadcrumbs/" }
```
Comment (vitvakatu) [inline app/gui/view/component-browser/component-list-panel/Cargo.toml]: Is it used in this crate?

File: app/gui/src/controller/searcher.rs [hunk 4]
```diff
@@ -374,11 +369,55 @@ impl Searcher {
...
+        let module = &component.defined_in;
+        let breadcrumbs_base = module.path().iter().map(|name| Breadcrumb::new_without_icon(name));
+        let breadcrumbs = breadcrumbs_base.chain(iter::once(Breadcrumb::new(name, icon)));
+        Ok(breadcrumbs.collect())
+    }
+
+    /// Return the documentation for the breadcrumb.
+    pub fn documentation_for_selected_breadcrumb(&self) -> Option<EntryDocumentation> {
+        let selected = self.breadcrumbs.selected();
+        console_log!("documentation_for_selected_breadcrumb -> selected: {:?}", selected);
```
Comment (vitvakatu) [inline app/gui/src/controller/searcher.rs]: `console_log` left in the code.

File: app/gui/src/controller/searcher.rs [hunk 5]
```diff
@@ -374,11 +369,55 @@ impl Searcher {
...
+            } else {
+                warn!(
+                    "Cannot update breadcrumbs with component that has no suggestion database \
+                entry. Invalid component: {:?}",
+                    component
+                );
+                None
+            }
+        } else {
+            warn!("Update readcrumbs called with invalid index: {}", index);
```
Comment (vitvakatu) [inline app/gui/src/controller/searcher.rs]:
```suggestion
            warn!("Update breadcrumbs called with invalid index: {}", index);
```

File: app/gui/src/controller/searcher/breadcrumbs.rs [hunk 6]
```diff
@@ -60,12 +63,13 @@ impl Breadcrumbs {
...
-        } else {
-            let index = self.selected.get();
-            self.list.borrow().get(index - 1).map(BreadcrumbEntry::id)
-        }
+        let index = self.selected.get();
+        self.list.borrow().get(index).map(BreadcrumbEntry::id)
+    }
+
+    /// Return a list of breadcrumbs to be displayed in the panel.
+    pub fn items(&self) -> impl Iterator<Item = BreadcrumbEntry> + '_ {
```
Comment (vitvakatu) [inline app/gui/src/controller/searcher/breadcrumbs.rs]: It looks suspicious lifetime-wise, and it seems to be unused.

File: app/gui/src/controller/searcher.rs [hunk 7]
```diff
@@ -374,11 +369,55 @@ impl Searcher {
...
+        }
+    }
+
+    /// Return the full breadcrumb for the entry.
+    pub fn breadcrumbs_for_entry(&self, id: EntryId) -> Result<Vec<Breadcrumb>, NoSuchEntry> {
+        let component = self.database.lookup(id)?;
+        let name = &component.name;
+        let icon = Some(component.icon());
+        let module = &component.defined_in;
+        let breadcrumbs_base = module.path().iter().map(|name| Breadcrumb::new_without_icon(name));
```
Comment (vitvakatu) [inline app/gui/src/controller/searcher.rs]: Here I see that we create breadcrumbs without icons for the modules path, but in the video, there are icons on each crumb. Is that correct?

File: lib/rust/ensogl/component/breadcrumbs/src/entry.rs (hunk 2 above)
Comment (MichaelMauderer) [reply lib/rust/ensogl/component/breadcrumbs/src/entry.rs]: Done.

File: app/gui/view/component-browser/component-list-panel/Cargo.toml (hunk 3 above)
Comment (MichaelMauderer) [reply app/gui/view/component-browser/component-list-panel/Cargo.toml]: No. Removed.

File: app/gui/src/controller/searcher/breadcrumbs.rs (hunk 6 above)
Comment (MichaelMauderer) [reply app/gui/src/controller/searcher/breadcrumbs.rs]: Removed.

File: app/gui/src/controller/searcher.rs (hunk 7 above)
Comment (MichaelMauderer) [reply app/gui/src/controller/searcher.rs]: This was left over and unused, but you are right that only the last breadcrumb should have an icon. Fixed.

**Review by vitvakatu — APPROVED** [review]
Code changes look good to me. @MichaelMauderer could you update the screencast in the PR description?

**Comment by MichaelMauderer** [comment]
> Code changes look good to me. @MichaelMauderer could you update the screencast in the PR description?

Updated.

**Comment by kazcw** [comment]
# QA: :yellow_circle:

When the documentation is hidden (with the Show/hide documentation button), the breadcrumb container remains:
image

Other than that, it looks good.

**Comment by farmaazon** [comment]
> When the documentation is hidden (with the Show/hide documentation button), the breadcrumb container remains:

This issue is also on develop, and I assume it will be fixed in the [documentation panel's update](7356). I would merge it.

**Comment by MichaelMauderer** [comment]
> > When the documentation is hidden (with the Show/hide documentation button), the breadcrumb container remains:
> 
> This issue is also on develop, and I assume it will be fixed in the [documentation panel's update](7356). I would merge it.

Yes, this is addressed in that PR. So I'll merge this one.

**Pull request merged**